README.md
Poor Maps
=========

![Not Maintained](https://img.shields.io/maintenance/no/2018.svg)

Poor Maps is an application for Sailfish OS to display tiled maps (e.g.
OpenStreetMap), places and routes with a flexible selection of data and
service providers.

Poor Maps is free software released under the GNU General Public License
(GPL), see the file [`COPYING`](COPYING) for details.

For testing purposes you can just run `qmlscene qml/poor-maps.qml`. For
installation, you can build the RPM package with command `make rpm`. You
don't need an SDK to build the RPM, only basic tools: `make`,
`rpmbuild`, `gettext` and `linguist` from `qttools`


NEWS.md
2018-04-02: Poor Maps 0.34.3
============================

* Remove Mapzen providers due to shutdown of services (#67)
* Revert back to an image canvas for the route polyline (#61)
* Update translations

2018-02-18: Poor Maps 0.34.2
============================

* Fix broken share position by SMS button
* Update translations

2018-01-06: Poor Maps 0.34.1
============================

* Fix colors of black icons to match overlays
* Update translations

2017-12-24: Poor Maps 0.34
==========================

* Add voice navigation (rinigus)
* Make Mapzen the default geocoder
* Improve route polyline rendering on HiDPI screens
* Make black map overlays less transparent
* Add donate button to the about page
* Update translations

2017-10-26: Poor Maps 0.33
==========================

* Add support for @3x tiles (#53)
* Add Cartago Streets basemaps @1x, @2x, @3x, @4x (#33)
* Set default basemap based on screen pixel ratio (#52)
* Allow navigation page button labels (begin, reroute, clear) to
  break across lines for better translatability
* Drop rerouting threshold to 100 meters
* Update translations

2017-08-26: Poor Maps 0.32
==========================

* Add support for rerouting (#26, rinigus)
* Redesign navigation status and narrative pages
* Add OSM Scout car day and night maps (rinigus)
* Allow searching by latitude/longitude coordinates, e.g.
  "60.169 24.941" or "60.169,24.941" (period as decimal, any
  non-alphanumeric separator between latitude and longitude)
* Fix auto-centering in landscape in navigation mode to not have
  the position icon overlap with the menu button (#46)
* Do relative requests to download tiles, fixes HSL map
* Fix string escaping in Python calls from QML (#50, rinigus)
* Add Russian translation
* Update translations

2017-07-09: Poor Maps 0.31
==========================

* Add city bikes to the Digitransit router
* Add ferry icon for navigation instructions
* Add hint texts to search screens
* Hide provider attribution behind long tap
* Update translations

2017-06-27: Poor Maps 0.30
==========================

* Merge OSM Scout routers (libosmscout and Valhalla)
* Fix language parameter for Mapzen/Valhalla routers

2017-06-15: Poor Maps 0.29
==========================

* Add support for OSM Scout Server's Valhalla router

2017-02-27: Poor Maps 0.28.4
============================

* Fix route polyline disappearing after minimizing Poor Maps (#35)
* Fix Digitransit router not returning anything if even one of the
  alternatives was fully a walking trip

2017-02-16: Poor Maps 0.28.3
============================

* Add Italian and Polish translations

2017-02-15: Poor Maps 0.28.2
============================

* Work around PyOtherSide's call_sync being broken in Sailfish OS
  2.1.0.9 "Iijoki"

2017-01-21: Poor Maps 0.28.1
============================

* Fix long routes not visible before moving or zooming the map
* Add German (Germany), Spanish (Spain), French, Dutch and Swedish
  translations

2017-01-04: Poor Maps 0.28
==========================

* Add user interface translations (see the file
  [`po/README.md`][0.28a] for details)

[0.28a]: https://github.com/otsaloma/poor-maps/blob/master/po/README.md

2016-12-13: Poor Maps 0.27
==========================

* Replace HSL geocoder and router with ones that use the new
  [Digitransit API][0.27a] (adds geocoding for whole Finland,
  real-time routing and experimental routing for Waltti and whole
  Finland regions, #19)
* Show route summary in the maneuver list page (#28)
* Don't show time remaining in the navigation instruction block
  to avoid overlapping text

[0.27a]: https://digitransit.fi/en/developers/

2016-11-26: Poor Maps 0.26.205
==============================

* Adapt to Nokia/HERE map plugin changes in Sailfish OS 2.0.5 "Haapajoki"
* Fix initial centering on new route sometimes being wrong

2016-11-01: Poor Maps 0.26
==========================

* Add [OSM Scout Server][0.26a] offline tiles, geocoder, nearby
  search and router – only listed in Poor Maps if you have OSM Scout
  Server installed and will only work if the server is running and
  you have made OpenStreetMap data available for the server (rinigus)
* Increase download thread count to match CPU core count for
  localhost (offline) tile servers
* Increase download timeout for localhost (offline) connections
* When navigating, make centering and auto-centering on position
  center the position on the part of the map visible below the
  navigation narrative, and further, if auto-rotate is on, center
  lower so that more map is shown ahead than behind (#14)
* When navigating, auto-rotate to match the route polyline instead
  of bearing calculated from GPS data, which should make the map
  rotate faster after a turn (#13)
* Fix smoothing of maneuver node icons along the route polyline
* Fix removing an item from search history to not show the removed
  item in the UI after refiltering

[0.26a]: https://openrepos.net/content/rinigus/osm-scout-server

2016-09-04: Poor Maps 0.25
==========================

* Fix padding, spacing, icon sizes and font sizes to work better
  across different size and different pixel density screens
* Move the menu button to the bottom center of screen
* Use [Mapbox's navigation icons][0.25b]
* Add scale filters to a pulldown menu in the basemap page
* Remove MapQuest Open tile source as tiles are [no longer available][0.25a]
* Add @2x variants of OpenCycleMap and Thunderforest Transport
* Include scale in all tile source names, including @1x
* Remove Nominatim geocoders
* Add [OpenCage](https://geocoder.opencagedata.com/) and
  [Photon](http://photon.komoot.de/) geocoders and make OpenCage the default
* Use the latest version of the OSRM router API
* Make Mapzen Turn-by-Turn the default router
* Indicate in lists of providers which is the default one
* New design for HSL Journey Planner results page
* Handle initial centering and zoom level better on first startup
  if positioning data not available

routers/osmscout.py
# -*- coding: utf-8 -*-

# Copyright (C) 2015 Osmo Salomaa
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
Routing using OSM Scout Server's Valhalla and libosmscout routers.

https://github.com/rinigus/osmscout-server
https://github.com/valhalla/valhalla-docs/blob/master/api-reference.md
"""

import copy
import json
import poor
import urllib.parse

CONF_DEFAULTS = {
    "type": "auto",
    "language": poor.util.get_default_language()
}

ICONS = {
     0: "flag",
     1: "depart",
     2: "depart-right",
     3: "depart-left",
     4: "arrive",
     5: "arrive-right",
     6: "arrive-left",
     7: "continue",
     8: "continue",
     9: "turn-slight-right",
    10: "turn-right",
    11: "turn-sharp-right",
    12: "uturn",
    13: "uturn",
    14: "turn-sharp-left",
    15: "turn-left",
    16: "turn-slight-left",
    17: "continue",
    18: "off-ramp-slight-right",
    19: "off-ramp-slight-left",
    20: "off-ramp-slight-right",
    21: "off-ramp-slight-left",
    22: "fork-straight",
    23: "fork-slight-right",
    24: "fork-slight-left",
    25: "merge-slight-left",
    26: "roundabout",
    27: "off-ramp-slight-right",
    28: "ferry",
    29: "depart",
    30: "flag",
    31: "flag",
    32: "flag",
    33: "flag",
    34: "flag",
    35: "flag",
    36: "flag",
}

ICONS_OSMSCOUT = {
    "destination": "arrive",
    "flag": "flag",
    "merge": "merge-slight-left",
    "motorway-change": "flag",
    "motorway-leave": "off-ramp-slight-right",
    "none": "flag",
    "roundabout-enter": "roundabout",
    "roundabout-exit": "off-ramp-slight-right",
    "start": "depart",
    "straight": "continue",
    "turn-left": "turn-left",
    "turn-right": "turn-right",
    "turn-sharp-left": "turn-sharp-left",
    "turn-sharp-right": "turn-sharp-right",
    "turn-slight-left": "turn-slight-left",
    "turn-slight-right": "turn-slight-right",
}

URL = "http://localhost:8553/v2/route?json={input}"
cache = {}

def prepare_endpoint(point):
    """Return `point` as a dictionary ready to be passed on to the router."""
    if isinstance(point, (list, tuple)):
        return dict(lat=point[1], lon=point[0])
    geocoder = poor.Geocoder("osmscout")
    results = geocoder.geocode(point, dict(limit=1))
    return prepare_endpoint((results[0]["x"], results[0]["y"]))

def route(fm, to, heading, params):
    """Find route and return its properties as a dictionary."""
    fm, to = map(prepare_endpoint, (fm, to))
    if heading is not None:
        fm["heading"] = heading
    if poor.conf.units == "american": units = "miles"
    elif poor.conf.units == "british": units = "miles"
    else: units = "kilometers"
    input = dict(locations=[fm, to],
                 costing=poor.conf.routers.osmscout.type,
                 directions_options=dict(language=poor.conf.routers.osmscout.language,
                                         units=units))

    input = urllib.parse.quote(json.dumps(input))
    url = URL.format(**locals())
    with poor.util.silent(KeyError):
        return copy.deepcopy(cache[url])
    result = poor.http.get_json(url)
    result = poor.AttrDict(result)
    if result.get("API version", "") == "libosmscout V1":
        return parse_result_libosmscout(url, result)
    return parse_result_valhalla(url, result)

def parse_result_libosmscout(url, result):
    """Parse and return route from libosmscout engine."""
    x, y = result.lng, result.lat
    maneuvers = [dict(
        x=float(maneuver.lng),
        y=float(maneuver.lat),
        icon=ICONS_OSMSCOUT.get(maneuver.get("type"), "flag"),
        narrative=maneuver.instruction,
        duration=float(maneuver.time),
        length=float(maneuver.length),
    ) for maneuver in result.maneuvers]
    route = dict(x=x, y=y, maneuvers=maneuvers)
    route["attribution"] = poor.util.get_routing_attribution(
        "OSM Scout", "libosmscout")
    route["language"] = result.language
    if route and route["x"]:
        cache[url] = copy.deepcopy(route)
    return route

def parse_result_valhalla(url, result):
    """Parse and return route from Valhalla engine."""
    legs = result.trip.legs[0]
    x, y = poor.util.decode_epl(legs.shape, precision=6)
    maneuvers = [dict(
        x=float(x[maneuver.begin_shape_index]),
        y=float(y[maneuver.begin_shape_index]),
        icon=ICONS.get(maneuver.type, "flag"),
        narrative=maneuver.instruction,
        verbal_alert=maneuver.get("verbal_transition_alert_instruction", None),
        verbal_pre=maneuver.get("verbal_pre_transition_instruction", None),
        verbal_post=maneuver.get("verbal_post_transition_instruction", None),
        duration=float(maneuver.time),
    ) for maneuver in legs.maneuvers]
    route = dict(x=x, y=y, maneuvers=maneuvers, mode="car")
    route["attribution"] = poor.util.get_routing_attribution(
        "OSM Scout", "Valhalla")
    route["language"] = result.trip.language
    if route and route["x"]:
        cache[url] = copy.deepcopy(route)
    return route
